### PR TITLE
[PRFC] Export kubernetes types and apirefs on path

### PR DESCRIPTION
**Pull request by szubster**
## Hey, I just made a Pull Request!

We are working on minimising initial bundle size of backstage.
We found xterm.js being very persistent on staying in initial js bundle, not the async one.

This PR is a PoC on potential attempt on fixing it, making react part of the plugin completely async loaded.

It refactors kubernetes plugin to more closely follow ADR11.
Additionally it creates a new path export in kubernetes-react. This path export should only contain apiRefs and interfaces. Those apiRefs and interfaces are then implemented by plugin package. They cannot live in main package, as components in react package use them.
Alternative approach is to extend ADR11 and introduce new package type, specially for types and apirefs, which will be used by both main and react package.
Another alternative is changing rollup config for `web-library` package type and either use [preserveModules](https://rollupjs.org/configuration-options/#output-preservemodules) (discouraged by rollup) or [plugin-multi-entry](https://www.npmjs.com/package/@rollup/plugin-multi-entry) to have path exports for web-libraries and leave interfaces and apiRefs in react packages.

Result on whole app bundle when this package is applied on our app:
image

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by szubster** [comment]
I am open for discussion what should be the best approach.
We use quite a few plugins which are not often used and would benefit from loading asynchronously.

File: plugins/kubernetes-react/package.json [hunk 1]
```diff
@@ -6,9 +6,22 @@
...
   "license": "Apache-2.0",
   "publishConfig": {
-    "access": "public",
-    "main": "dist/index.esm.js",
-    "types": "dist/index.d.ts"
+    "access": "public"
+  },
+  "exports": {
+    ".": "./src/index.ts",
+    "./api": "./src/api/index.ts",
```
Comment (szubster) [inline plugins/kubernetes-react/package.json:13]: This is the most important part.
When this path export is used in `kubernetes` package, it allows only the APIs be loaded synchronously, and all the react stuff is dynamically loaded when kubernetes plugin is navigated to.

**Comment by szubster** [comment]
@freben What do you think?
Is there an approach you like more from listed ones, and how can I proceed?
Getting it done repo-wide would be a lot of effort, but we can start small.

**Comment by szubster** [comment]
@freben Bumping :)

**Comment by mclarke47** [comment]
Sorry for the slow reply, will look this week

File: plugins/kubernetes/src/index.ts [hunk 2]
```diff
@@ -24,8 +24,12 @@ export {
   kubernetesPlugin,
   kubernetesPlugin as plugin,
   EntityKubernetesContent,
+  EntityKubernetesContent as Router,
 } from './plugin';
-export type { EntityKubernetesContentProps } from './plugin';
-export { Router, isKubernetesAvailable } from './Router';
-// TODO remove this re-export as a breaking change after a couple of releases
-export * from '@backstage/plugin-kubernetes-react';
```
Comment (mclarke47) [inline plugins/kubernetes/src/index.ts:31]: I noticed we removed the re-export. I think we should leave this in for a little longer

**Comment by mclarke47** [comment]
Are there any backwards incompatible changes that need to be documented?

**Comment by szubster** [comment]
There are. It's a PRFC, as I wanted to check what would be general recommendation.
Maye even ADR11 should be extended to add new package type? That package type would only contain apirefs and types. How to call them? I do not know...

Currently apirefs reside in `x-react` packages. And `x-react` packages generally produce a single index.js file in dist. This causes importing such apiref (and `x` needs to be loaded synchronously), importing whole package. In the end, lazily exporting components from `x` is not really working.
Adding this new package type would decouple sync part of `x` (setting up apis) from async part of `x` (react code).
apirefs cannot themself live in `x` package, as `x-react` needs those apis.

Or the rollup config amended for all packages, so that importing tree-shaking and importing single function/class does not load whole plugin.

If one of those solutions is accepted, either I will try to amend rollup config (I think this should be possible to make backward-compatible) or create new PR with new ADR, that will supersed ADR11, extending it by new package type.

**Pull request closed without merging**